src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -58,48 +61,122 @@
+     * This method is to provide a replacement for InetSocketAddress.getHostString().
+     *
+     * @param addr
+     * @return Hostname string of address parameter
+     */
+    private String getHostString(InetSocketAddress addr) {
+        String hostString = "";
+
+        if (addr == null) {
+            return hostString;
+        }
+        if (!addr.isUnresolved()) {
+            InetAddress ia = addr.getAddress();
+
+            // If the string starts with '/', then it has no hostname
+            // and we want to avoid the reverse lookup, so we return
+            // the string representation of the address.
+            if (ia.toString().startsWith("/")) {
+                hostString = ia.getHostAddress();
+            } else {
+                hostString = addr.getHostName();
+            }
+        } else {
+            // According to the Java 6 documentation, if the hostname is
+            // unresolved, then the string before the colon is the hostname.
+            String addrString = addr.toString();
+            hostString = addrString.substring(0, addrString.lastIndexOf(':'));
+        }
+
+        return hostString;
+    }
+
     public int size() {
         return serverAddresses.size();
     }
 
+    // Counts the number of addresses added and removed during
+    // the last call to next. Used mainly for test purposes.
+    // See StasticHostProviderTest.
+    private int nextAdded = 0;
+    private int nextRemoved = 0;
+
+    public int getNextAdded() {
+        return nextAdded;
+    }
+
+    public int getNextRemoved() {
+        return nextRemoved;
+    }
+
     public InetSocketAddress next(long spinDelay) {
-        ++currentIndex;
-        if (currentIndex == serverAddresses.size()) {
-            currentIndex = 0;
+        // Handle possible connection error by re-resolving hostname if possible
+        if (!connectedSinceNext) {
+            InetSocketAddress curAddr = serverAddresses.get(currentIndex);
+            String curHostString = getHostString(curAddr);
+            if (!curHostString.equals(curAddr.getAddress().getHostAddress())) {
+                LOG.info("Resolving again hostname: {}", getHostString(curAddr));
+                try {
+                    int thePort = curAddr.getPort();
+                    InetAddress resolvedAddresses[] = InetAddress.getAllByName(curHostString);
+                    nextAdded = 0;
+                    nextRemoved = 0;
+                    if (resolvedAddresses.length == 1) {
+                        serverAddresses.set(currentIndex, new InetSocketAddress(resolvedAddresses[0], thePort));
+                        nextAdded = nextRemoved = 1;
+                        LOG.debug("Newly resolved address: {}", resolvedAddresses[0]);
+                    } else {
+                        int i = 0;
+                        while (i < serverAddresses.size()) {
+                            if (getHostString(serverAddresses.get(i)).equals(curHostString) &&
```

Comment:
as i mentioned in https://github.com/apache/zookeeper/pull/150/files#r98543324 this all gets a little complicated? What do you think about using a map to trap all these associations?

Reply:
It seems that this change makes things a little a bit even more complicated, because `next()` method needs to have a list of IP addresses to iterate on. Maintenance is easier with a Map, but I need to find a way to use it as a list.

Reply:
How to shuffle the elements?

Reply:
You could put the `keySet` in a list and shuffle?

Reply:
The key is the hoststring which is like `issues.apache.org` and could have multiple InetAddresses which are the values of the key in the map. If I shuffle only the keys, I can iterate on only the keys and not sure how to choose from the addresses.

Reply:
Why don't we simplify things and just do this always? In other words, always remove all serverAddresses that correspond to the curHostString. That way if curHostString goes from resolving to many addresses to just 1, we remove all of them?

Reply:
That's exactly the point. 
(think I've already suggested the same thing in an outdated post)